packages/repository/src/relations/relation.types.ts

```diff
@@ -108,3 +110,25 @@ export type RelationMetadata =
 
 // Re-export Getter so that users don't have to import from @loopback/context
 export {Getter} from '@loopback/context';
+
+/**
+ * @returns An array of resolved values, the items must be ordered in the same
+ * way as `sourceEntities`. The resolved value can be one of:
+ * - `undefined` when no target model(s) were found
+ * - `Entity` for relations targeting a single model
+ * - `Entity[]` for relations targeting multiple models
+ */
+export type InclusionResolver<S extends Entity, T extends Entity> = (
+  /**
+   * List of source models as returned by the first database query.
+   */
+  sourceEntities: S[],
```

Comment:
I wonder if `sourceEntities` needs to have a generic type `S`. It's not helping the compiler much.

Reply:
@raymondfeng I suggested to add a generic type `S` to avoid specifying the source entity's type like

```ts
const resolver: InclusionResolver = async entities => {
        // you need to specify Category....
        const c = entities[0] as Category;
        const product: Product[] = await categoryRepo.products(c.id).find();
        return [product];
      };

      categoryRepo.inclusionResolvers.set('products', resolver);

```

In the resolver function.

See the current [resolver function](https://github.com/strongloop/loopback-next/blob/02fe72714e56e035495ef37fb72ab18922f15ca7/packages/repository/src/__tests__/unit/repositories/relation.helpers.unit.ts#L321-L332)

vs

[previous resolver function](https://github.com/strongloop/loopback-next/blob/5e5ae8d8ccedfeea0d4c4cf7017ad5be88f45c6f/packages/repository/src/__tests__/unit/repositories/relation.helpers.unit.ts#L178-L188)

Reply:
I see. It sounds good then.

Reply:
On a second thought...maybe the resolver should be written as
```ts
const hasManyResolver: InclusionResolver = async (entities: Category[]) => {
    const products: Product[] = [];

    for (const category of entities) {
      const product = await categoryRepo.products(category.id).find();
      products.push(product);
    }
    return products;
  };
```
?

Reply:
The current signature with `S` seems to be better.

Reply:
IIRC, generic arguments are pretty useless here. At high level, we need a registry of resolvers, mapping from relation name (e.g. `products`) to a resolver. This registry cannot use generic types and has to fall back to `Entity`.

That's why I am using the following definition in my spike:

https://github.com/strongloop/loopback-next/blob/2fa5df67181cdcd23a5dce90c9c640fe75943cb8/packages/repository/src/relations/relation.types.ts#L121-L134

I am concerned that the current design creates a false sense of safety - the resolver is implemented with the assumption that it will be called with certain `S` and `T` types, but in reality the code calling it will always allow any `Entity`.

Having said that, I see how the stubbed resolvers are easier to implement when generic arguments are in place.

I don't have enough insight now to be able to judge what aspect is more important.

Reply:
Ha, I take my comment back. When I reviewed code below, I found that is able to provide a meaningful value for `S`.

```ts
public readonly inclusionResolvers: Map<
  string,
  InclusionResolver<T, Entity>
> = new Map();
```

Nice trick! I agree that we should keep the generic argument for the type of the source entity.

The remaining question is whether we need the second generic argument for the type of the target (related) entity?

Reply:
It would help the developer with the return type for each resolver (e.g. restricting to type `Product`), but I don't think it's too important.